rslib/src/scheduler/fsrs/memory_state.rs

```diff
@@ -79,124 +92,247 @@ impl Collection {
-                                            } else {
-                                                &mut card.due
-                                            };
-                                            let new_due = (timing.days_elapsed as i32)
-                                                - days_elapsed
-                                                + card.interval as i32;
-                                            if let Some(rescheduler) = &mut rescheduler {
-                                                rescheduler.update_due_cnt_per_day(
-                                                    *due,
-                                                    new_due,
-                                                    deckconfig_id,
-                                                );
-                                            }
-                                            *due = new_due;
-                                            // Add a rescheduled revlog entry
-                                            self.log_rescheduled_review(
-                                                &card,
-                                                original_interval,
-                                                usn,
-                                            )?;
-                                        }
-                                    }
-                                }
-                            }
-                        }
+                        Either::Left((card_id, item))
                     } else {
-                        // clear memory states if item is None
-                        card.memory_state = None;
+                        Either::Right(card_id)
                     }
+                });
+
+            let decay = get_decay_from_params(&req.params);
+
+            // Store decay and desired retention in the card so that add-ons, card info,
+            // stats and browser search/sorts don't need to access the deck config.
+            // Unlike memory states, scheduler doesn't use decay and dr stored in the card.
+            let set_decay_and_desired_retention = move |card: &mut Card| {
+                let deck_id = card.original_or_current_deck_id();
+
+                let desired_retention = *req
+                    .deck_desired_retention
+                    .get(&deck_id)
+                    .unwrap_or(&req.preset_desired_retention);
+
+                card.desired_retention = Some(desired_retention);
+                card.decay = Some(decay);
+            };
+
+            self.update_memory_state_for_itemless_cards(
+                cards_without_items,
+                set_decay_and_desired_retention,
+                usn,
+                &mut on_updated_card,
+            )?;
+
+            let mut rescheduler = self
+                .get_config_bool(BoolKey::LoadBalancerEnabled)
+                .then(|| Rescheduler::new(self))
+                .transpose()?;
```

Comment:
```suggestion
            let mut rescheduler = req
                .reschedule
                .bitand(self.get_config_bool(BoolKey::LoadBalancerEnabled))
                .then(|| Rescheduler::new(self))
                .transpose()?;
```

Reply:
We can skip `Rescheduler::new(self)` if `req.reschedule` is `false`. It could save ~60ms in a large collection.